### ISSUE TITLE
Circular reference misdetection (introduced after 3.6.3)

### ISSUE THREAD

**Issue by orzFly** [microsoft/TypeScript#34960]
**TypeScript Version:**  3.7.2

**Search Terms:** 7022, referenced directly or indirectly in its own initializer

**Code**

```ts
let items: string[] | null = null;
items = items || [];

for (const row of []) {
  const item = items.find(() => true); // ERROR
  const result = item ? true: false    // ERROR

  switch (result) {

  }
}
```

**Expected behavior:**
No error & warnings.
```
const item: string | undefined
const result: boolean
```

**Actual behavior:**
```
'item' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)
'result' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)
```

```
const item: any
const result: any
```

**Playground Link:** [Link](https://www.typescriptlang.org/play/index.html#code/FAGwpgLgBAlhYFsDOAuKSICcYDsDmA2gLpQA+UOAriCFALwXUgDcwciS9s8yZ5xrYADMA9pigAKAMYicGKJhEB3KCKFRiASigBvYFCgy50dgi6mkAOiG4AJhIna6APihZKYTawNH5mMEjU0AymUAD8bpgeaEIAhiBIYMD66EpwUgAWkv6BIBDaeikAvsAlQA)

It works in 3.6.3 and fails in 3.7.2.

**Related Issues:** 
I think this is not related to #33191 since this introduced after 3.6.3.

**Comment by orzFly** [comment]
More simplified:

[Playground link](https://www.typescriptlang.org/play/index.html#code/MYewdgzgLgBAZgGwIYHMBcMBGIQIKZJgwC88SCEeA3AFA1wgBOMAFKJLIyAO4whwwA2gF0AlDADeNGDHbQYjPBACuCWKUSoYAfhhRGyvBjjlK0mBG4BLKMAAWrRSrXiJMAL41PQA)

```typescript
const flag: boolean = false;

for (const row of []) {
  const result = flag ? true: false
  switch (result) { }
}
```

Expected: 
```
const result: boolean
```

Actual: 
```
const result: any
'result' implicitly has type 'any' because it does not have a type annotation and is referenced directly or indirectly in its own initializer.(7022)
```

**Comment by ahejlsberg** [comment]
This is fixed by #35000.
